sql-plugin/src/main/scala/org/apache/spark/sql/rapids/urlFunctions.scala

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.rapids
+
+import java.net.URISyntaxException
+
+import ai.rapids.cudf.{ColumnVector, DType, RegexProgram, Scalar, Table}
+import com.nvidia.spark.rapids._
+import com.nvidia.spark.rapids.Arm._
+import com.nvidia.spark.rapids.RapidsPluginImplicits._
+import com.nvidia.spark.rapids.shims.ShimExpression
+
+import org.apache.spark.sql.catalyst.analysis.TypeCheckResult
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.rapids.shims.RapidsErrorUtils
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.vectorized.ColumnarBatch
+import org.apache.spark.unsafe.types.UTF8String
+
+object GpuParseUrl {
+  private val HOST = "HOST"
+  private val PATH = "PATH"
+  private val QUERY = "QUERY"
+  private val REF = "REF"
+  private val PROTOCOL = "PROTOCOL"
+  private val FILE = "FILE"
+  private val AUTHORITY = "AUTHORITY"
+  private val USERINFO = "USERINFO"
+  private val REGEXPREFIX = """(&|^|\?)"""
+  private val REGEXSUBFIX = "=([^&]*)"
+}
+
+case class GpuParseUrl(children: Seq[Expression], 
+    failOnErrorOverride: Boolean = SQLConf.get.ansiEnabled) 
+  extends GpuExpression with ShimExpression with ExpectsInputTypes {
+
+  def this(children: Seq[Expression]) = this(children, SQLConf.get.ansiEnabled)
+
+  override def nullable: Boolean = true
+  override def inputTypes: Seq[DataType] = Seq.fill(children.size)(StringType)
+  override def dataType: DataType = StringType
+  override def prettyName: String = "parse_url"
+
+  import GpuParseUrl._
+  
+  override def checkInputDataTypes(): TypeCheckResult = {
+    if (children.size > 3 || children.size < 2) {
+      RapidsErrorUtils.parseUrlWrongNumArgs(children.size) match {
+        case res: Some[TypeCheckResult] => return res.get
+        case _ => // error message has been thrown
+      }
+    }
+    super[ExpectsInputTypes].checkInputDataTypes()
+  }
+
+  private def getPattern(key: UTF8String): RegexProgram = {
+    val regex = REGEXPREFIX + key.toString + REGEXSUBFIX
+    new RegexProgram(regex)
+  }
+
+  private def reValid(url: ColumnVector): ColumnVector = {
+    // TODO: Validite the url
+    val regex = """^[^ ]*$"""
```

Comment:
If all we are doing is checking to see if the URL has white space in it, there are faster ways to do that. `stringContains` is the most obvious example I can think of.

Reply:
After I fixed all cases I know, the only thing I want to check before extracting is still checking  if the URL has white space in it. Updated to stringContains.